src/schemes/fluid/surface_normal_sph.jl

```diff
@@ -0,0 +1,144 @@
+@doc raw"""
+    ColorfieldSurfaceNormal()
+
+Color field based computation of the interface normals.
+"""
+struct ColorfieldSurfaceNormal end
+
+function create_cache_surface_normal(surface_normal_method, ELTYPE, NDIMS, nparticles)
+    return (;)
+end
+
+function create_cache_surface_normal(::ColorfieldSurfaceNormal, ELTYPE, NDIMS, nparticles)
+    surface_normal = Array{ELTYPE, 2}(undef, NDIMS, nparticles)
+    neighbor_count = Array{ELTYPE, 1}(undef, nparticles)
+    return (; surface_normal, neighbor_count)
+end
+
+@inline function surface_normal(particle_system::FluidSystem, particle)
+    (; cache) = particle_system
+    return extract_svector(cache.surface_normal, particle_system, particle)
+end
+
+# Section 2.2 in Akinci et al. 2013 "Versatile Surface Tension and Adhesion for SPH Fluids"
+# Note: This is the simplest form of normal approximation commonly used in SPH and comes
+# with serious deficits in accuracy especially at corners, small neighborhoods and boundaries
+function calc_normal_akinci!(system, neighbor_system::FluidSystem, u_system, v,
+                             v_neighbor_system, u_neighbor_system, semi, surfn)
+    (; cache) = system
+
+    system_coords = current_coordinates(u_system, system)
+    neighbor_system_coords = current_coordinates(u_neighbor_system, neighbor_system)
+    nhs = get_neighborhood_search(system, neighbor_system, semi)
+
+    foreach_point_neighbor(system, neighbor_system,
+                           system_coords, neighbor_system_coords,
+                           nhs) do particle, neighbor, pos_diff, distance
+        m_b = hydrodynamic_mass(neighbor_system, neighbor)
+        density_neighbor = particle_density(v_neighbor_system,
+                                            neighbor_system, neighbor)
+        grad_kernel = smoothing_kernel_grad(system, pos_diff, distance)
+        for i in 1:ndims(system)
+            cache.surface_normal[i, particle] += m_b / density_neighbor * grad_kernel[i]
+        end
+
+        cache.neighbor_count[particle] += 1
+    end
+
+    return system
+end
+
+# Section 2.2 in Akinci et al. 2013 "Versatile Surface Tension and Adhesion for SPH Fluids"
+# Note: This is the simplest form of normal approximation commonly used in SPH and comes
+# with serious deficits in accuracy especially at corners, small neighborhoods and boundaries
+function calc_normal_akinci!(system, neighbor_system::BoundarySystem, u_system, v,
```

Comment:
Please check Codecov report. This is missing tests.

Reply:
Testing this doesn't make much sense. Since there is no analytical value. And it will change in all subsequent PRs.

Reply:
Let me restate that there is a analytical value but the results will not be close. Just closer than without it.

Reply:
There must be some useful way to test this. Maybe with a high resolution and very large smoothing length?